### ISSUE TITLE
Trying to exclude certain methods from protection

### ISSUE THREAD

**Issue by dspwhite** [yck1509/ConfuserEx#393]
From my C# project, I want to allow control flow obfuscation for all of it, except a single method or class. I am unable to do this.

------------- First attempt
I tried adding the following line above my Form partial class:
`[Obfuscation(Exclude = true, Feature = "ctrl flow")]`

...and even this line above the namespace in that same file:
`[assembly: Obfuscation(Exclude = true, Feature = "ctrl flow")]`

Then from the ConfuserEx 0.6 GUI, I saved a project with "Control Flow protection" added (with Preset: None). Then from the CLI, I executed:

`Confuser.CLI.exe R:\.crproj`

Unfortunately, the exe was still control flow protected, evident by the slow running of the method/class which is supposed to exclude control flow protection.

-------------- Second attempt
Then I tried:
`Confuser.CLI.exe -o="R:\output" myprog.exe my.dll`

In this situation, because I'm not using a crproj file, I am unable to get any protection added at all. I tried this to no avail (-ID is not a valid CLI parameter by the looks of it):
`Confuser.CLI.exe -o="R:\output" myprog.exe my.dll` -ID="ctrl flow"

What am I missing?

**Comment by yck1509** [comment]
Hi,
You may want to try the latest build from the [CI Server](https://ci.appveyor.com/project/yck1509/ConfuserEx). The version 0.6 does not supports mixed usage of crproj and ObfuscationAttribute.

**Comment by dspwhite** [comment]
Oh will that feature make its way into 0.7? I'd rather use the official version and wait for that if possible.

**Comment by dspwhite** [comment]
Okay downloaded the cutting edge version, and tried putting:

> [Obfuscation(Exclude = true, Feature = "ctrl flow")]

...above a class in the code. I'm getting the error:

> Exception: System.ArgumentException: Feature property cannot be set when Exclude is true.

Any thoughts?

**Comment by Drizin** [comment]
I had the same problem. Altough the exception message is very straightforward, I had to study parts of the code and documentation to understand what was happening: **You should never use both `Exclude=true` (which unfortunately is the default) \*simultaneously\* with some other `Feature="xxx"`**:
- If you add `Exclude=true` in some item (please see [hierarchy](https://github.com/yck1509/ConfuserEx/wiki/Declarative-Obfuscation)), the item is completely excluded from obfuscation. 
- If you want to give some special instruction(s) to the obfuscator (like the common "don't rename types because this will be serialized"), then you should indicate those instructions in the `Feature` but you should explicitly set `Exclude=false`. 

This follows the [documentation for the attribute](https://msdn.microsoft.com/en-us/library/system.reflection.obfuscationattribute(v=vs.110).aspx), altough other obfuscators use the `Exclude` for different things, not following the official documentation. ConfuserEx uses a declarative syntax different from other obfuscators, but it has a very powerful syntax (altough I think it could have better documentation). 

BTW: If it helps someone, while for most other obfuscators you should use `Feature="renaming"` to indicate that something shouldn't be renamed, in ConfuserEx the correct syntax is `Feature="-rename"`.

**Comment by dspwhite** [comment]
Thanks for your reply. I'm not sure I follow. I've changed from `[Obfuscation(Exclude = true, Feature = "ctrl flow")]` to just: `[Obfuscation(Feature = "ctrl flow")]` (assuming exclude being true is the default), and I still get the error.

Could you be more explicit with how I should fix this. Are you saying something like that instead of excluding a single particular method, I have to 'include' ALL of my dozens/hundreds of other methods using `[Obfuscation(Exclude = false, Feature = "ctrl flow")]`, and putting that statement above ALL (except the one I want to exclude) of the methods in the entire project? If so, that sounds very non-intuitive and cumbersome.

**Comment by Drizin** [comment]
Sorry, I guess I was not clear. I just explained that adding `Exclude=true` should **not** be used, because that's supposed to completely **remove** some object from obfuscation - and if you only want to change the obfuscation behavior (like disable control flow, in your case) then you should just add/remove the individual instructions using the Feature attribute (using [+ and - syntax](https://github.com/yck1509/ConfuserEx/wiki/Declarative-Obfuscation) to add/remove obfuscation steps). However, I forgot to emphasize that `Exclude` is `true` by default. So in this, case you should **explicitly** set it as false.

Concrete example for your case. This is an **overriding rule** in the sense that it overrides your general obfuscation rules:
````csharp
// this class should NOT be excluded from obfuscation, but control flow feature will be disabled here
[Obfuscation(Exclude = true, Feature = "-ctrl flow")]
public class YourClass {
...
}
````

The **general obfuscation rules** (which we can override on a class level or member level, as example above) can be defined either in your obfuscation project (as you did) but can also be declared in `AssemblyInfo.cs`. 
````csharp
[assembly: Obfuscation(Exclude = false, Feature = "random seed: MySeed123")]
[assembly: Obfuscation(Exclude = false, Feature = "packer:compressor")]
[assembly: Obfuscation(Exclude = false, Feature = "preset(normal)")]
````
Please note that [global assembly-level attributes](https://msdn.microsoft.com/en-us/library/284c1c4s(v=vs.90).aspx) do not belong to a single namespace, even if you declare them right above your namespace - they are global indeed. Keeping them inside `AssemblyInfo.cs` is just a convention.

In your specific case, since you defined in the obfuscation project the **Preset: None** and added **Control Flow Protection**, you don't need any assembly-level attributes. Just need to declare the exceptions (overriding rules) right above your classes (or members) like the first code block.

If you'd rather keep the global obfuscation rules in the source code (and in this case you wouldn't need to configure control flow protection in the `.crproj`), you can use something like this:

````csharp
[assembly: Obfuscation(Exclude = false, Feature = "preset(none);+ctrl flow;")]
````
If you also want to keep your obfuscation exceptions (overriding rules) as assembly-level attributes (instead of applying them directly to the classes), you can use something like this (I haven't used this, not sure if my syntax/understanding is correct):
````csharp
[assembly: Obfuscation(Exclude = false, Feature = "namespace 'MyNameSpace':-ctrl flow")]
// or
[assembly: Obfuscation(Exclude = false, Feature = "match('^MyNameSpace.') and member-type('type'):-ctrl flow")]
````

As you can see, the syntax is powerful (you can even add [priorities](https://github.com/yck1509/ConfuserEx/issues/374) so that some rule can override the other).
I prefer to **keep obfuscation rules directly applied above the classes/members** (not on assembly level, let alone in the `.crproj`) because it **makes refactoring easier**: if you move some class to another namespace, your obfuscation instructions (class attributes) will always follow your class.   
And since I declare my obfuscation exceptions directly into the code, [I believe it makes sense to declare the general obfuscation rules also in the code (and not in the obfuscation project)](http://drizin.com.br/how-to-use-confuserex-obfuscator/) - that's why I like the `AssemblyInfo.cs` approach (or sometimes I call it `ObfuscationSettings.cs`).

**Comment by dspwhite** [comment]
Aha, it works! Thank you. The thing I missed was that the 'Exclude' keyword applies to ANY type of obfuscation, and I missed that to omit the Control Flow type I needed a minus symbol in front of it (i.e: `[Obfuscation(Exclude = false, Feature = "-ctrl flow")]`

Btw, I think there's a typo in your first snippet which should say `Exclude = false` rather than true.

**Comment by rolandh** [comment]
The the you need a minus or plus symbol is not clearly mentioned in the documentation, it was only after reading this I understood it.

**Comment by Drizin** [comment]
@rolandh I'm glad it helped. I also couldn't find any documentation about this, and I had to dig deep into source code and make dozens of tests to understand how it works.

**Comment by rolandh** [comment]
Is the documentation publicly editable? You should paste in your writeup. It would save a lot of people time.

**Comment by IllarionRUS** [comment]
Hey, too, can not turn off obfuscation, found that the problem is to rename, cut off the class renaming "-", it is still not working. `[Obfuscation(Exclude = false, Feature = "-rename")]`
'System.Data.Entity.ModelConfiguration.ModelValidationException' occurred in Unknown Module. 
Additional information: During the creation of the model validation errors found. EcigTools.DbDevices:: To EntityType "DbDevices" keys are not defined. Identify key for this EntityType. If you have thoughts prompt.

**Comment by rolandh** [comment]
That error indicates you have a different problem. I would start a new issue.
